crates/iceberg/src/scan.rs

```diff
@@ -158,8 +196,24 @@ impl TableScan {
             .await?;
 
             // Generate data file stream
-            let mut entries = iter(manifest_list.entries());
-            while let Some(entry) = entries.next().await {
+            for entry in manifest_list.entries() {
+                // If this scan has a filter, check the partition evaluator cache for an existing
+                // PartitionEvaluator that matches this manifest's partition spec ID.
+                // Use one from the cache if there is one. If not, create one, put it in
+                // the cache, and take a reference to it.
+                #[allow(clippy::map_entry)]
+                if let Some(filter) = filter.as_ref() {
+                    if !manifest_evaluator_cache.contains_key(&entry.partition_spec_id) {
```

Comment:
TBH I prefer the entry api which make code more concise, but this is not a blocker.

Reply:
I also prefer the entry API, but `Self::create_manifest_evaluator` returns a `Result<ManifestEvaluator>` rather than a `ManifestEvaluator`. The `?` operator can't be used inside the closure passed to the Entry API's [or_insert_with_key](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert_with_key) method, and this was the least ugly way I found of handling that. If you know of a nice way of using the `or_insert_with_key` with en entry_generating function that returns a `Result`, please share! :-)

Reply:
Oh, got it. Thanks for clarification.